components/tools/OmeroWeb/omeroweb/webgateway/views.py

```diff
@@ -1212,31 +1223,46 @@ def wrap(request, *args, **kwargs):
         try:
             server_id = kwargs.get('server_id', None)
             if server_id is None:
-                server_id = request.session['connector'].server_id
-            kwargs['server_id'] = server_id
+                if 'connector' in request.session:
+                    server_id = request.session['connector'].server_id
+            if server_id is not None:
+                kwargs['server_id'] = server_id
             rv = f(request, *args, **kwargs)
             if kwargs.get('_raw', False):
                 return rv
             if isinstance(rv, HttpResponse):
                 return rv
-            rv = json.dumps(rv)
             c = request.GET.get('callback', None)
             if c is not None and not kwargs.get('_internal', False):
+                rv = json.dumps(rv)
                 rv = '%s(%s)' % (c, rv)
+                # mimetype for JSONP is application/javascript
+                return HttpJavascriptResponse(rv)
             if kwargs.get('_internal', False):
                 return rv
-            return HttpJavascriptResponse(rv)
-        except omero.ServerError:
+            # mimetype for JSON is application/json
+            # NB: To support old api E.g. /get_rois_json/
+            # We need to support lists
+            # TODO: Have /api/ not use @jsonp.
+            safe = type(rv) is dict
+            return JsonResponse(rv, safe=safe)
+        except Exception, ex:
+            # Default status is 500 'server error'
+            # But we try to handle all 'expected' errors appropriately
+            status = 500
+            if isinstance(ex, omero.SecurityViolation):
```

Comment:
Yes, I already had to move ```ServerException``` last (when I needed it because default was 400). Now I've removed it and replaced ```ValidationException`` with ```ApiUsageException``` in https://github.com/openmicroscopy/openmicroscopy/pull/4708/commits/a4d678c9c4708e1fa2e25d6ecac8a488f8d76e6a